File: TotalCrossVM/src/nm/crypto/MD5Digest.c
// Copyright (C) 2000-2013 SuperWaba Ltda.
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only



#include "tcvm.h"
#include "../../axtls/crypto.h"

//////////////////////////////////////////////////////////////////////////
TC_API void tcdMD5D_init(NMParams p) // totalcross/crypto/digest/MD5Digest native private void init();
{
   TCObject md5dObj = p->obj[0];
   TCObject digestObj;

   if ((digestObj = createByteArray(p->currentContext, sizeof(MD5_CTX))) != null)
   {
      *Digest_digestRef(md5dObj) = digestObj;
      setObjectLock(digestObj, UNLOCKED);
   }
}
//////////////////////////////////////////////////////////////////////////
TC_API void tcdMD5D_process_B(NMParams p) // totalcross/crypto/digest/MD5Digest native protected final byte[] process(byte []data);
{
   TCObject md5dObj = p->obj[0];
   TCObject dataObj = p->obj[1];
   TCObject digestObj = *Digest_digestRef(md5dObj);
   MD5_CTX *ctx = (MD5_CTX*) ARRAYOBJ_START(digestObj);

   TCObject byteArrayResult;

   if (!dataObj)
   {
      throwNullArgumentException(p->currentContext, "data");
      return;
   }

   MD5_Init(ctx);
   MD5_Update(ctx, ARRAYOBJ_START(dataObj), ARRAYOBJ_LEN(dataObj));

   if ((byteArrayResult = createByteArray(p->currentContext, MD5_SIZE)) != null)
   {
      p->retO = byteArrayResult;
      setObjectLock(byteArrayResult, UNLOCKED);
      MD5_Final(ARRAYOBJ_START(byteArrayResult), ctx);
   }
}

File: TotalCrossSDK/src/totalcross/crypto/digest/MD5Digest.java
/*********************************************************************************
 *  TotalCross Software Development Kit                                          *
 *  Copyright (C) 2000-2012 SuperWaba Ltda.                                      *
 *  All Rights Reserved                                                          *
 *                                                                               *
 *  This library and virtual machine is distributed in the hope that it will     *
 *  be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of    *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.                         *
 *                                                                               *
 *  This file is covered by the GNU LESSER GENERAL PUBLIC LICENSE VERSION 3.0    *
 *  A copy of this license is located in file license.txt at the root of this    *
 *  SDK or can be downloaded here:                                               *
 *  http://www.gnu.org/licenses/lgpl-3.0.txt                                     *
 *                                                                               *
 *********************************************************************************/

package totalcross.crypto.digest;

import java.security.MessageDigest;
import totalcross.crypto.*;

/**
 * This class implements the MD5 message digest algorithm.
 */
public class MD5Digest extends Digest
{
   /**
    * Creates a new MD5Digest object.
    * 
    * @throws NoSuchAlgorithmException If no Provider supports a <code>MessageDigestSpi</code> implementation for the specified algorithm.
    */
   public MD5Digest() throws NoSuchAlgorithmException
   {
      try
      {
         digestRef = MessageDigest.getInstance("MD5");
      }
      catch (java.security.NoSuchAlgorithmException e) 
      {
         throw new NoSuchAlgorithmException(e.getMessage());
      }
   }
   
   /**
    * Returns the name of the algorithm.
    * 
    * @return "MD5".
    */
   public final String getAlgorithm()
   {
      return "MD5";
   }
   
   /**
    * Returns the block length.
    * 
    * @return 64.
    */
   public final int getBlockLength()
   {
      return 64;
   }
   
   /**
    * Returns the message digest length.
    * 
    * @return 16.
    */
   public final int getDigestLength()
   {
      return 16;
   }
   
   protected final byte[] process(byte[] data)
   {
      MessageDigest digest = (MessageDigest)digestRef;
      digest.reset();
      digest.update(data);
      
      return ((MessageDigest)digestRef).digest();
   }
}
